### ISSUE TITLE
10 k limit

### ISSUE THREAD

**Issue by joemoe** [raintank/crate-datasource#11]
at the moment we hit the crate internal 10 k limit when want to visualize a table grouping by a field and going for seconds over a longer time span like an hour or something.

the graph then stops at one point.
do you have an idea what we can do about this? e.g.:
* run several queries to complete the graph.
* automatically change the resolution when we know the cardinality of the group by column.

<img width="1263" alt="bildschirmfoto 2016-08-04 um 17 38 55" src="https://cloud.githubusercontent.com/assets/68620/17408238/b791812e-5a6a-11e6-8559-66566bacd6ad.png">

**Comment by alexanderzobnin** [comment]
The main problem is that Grafana don't knows how many points will be returned. So we can check it only after handling request.

**Comment by joemoe** [comment]
I was thinking on the resolution. E.g. if we'd know we got a timespan from 11 to 17:30 which are 6 hours 30 minutes or ~ 23.000 seconds. it wouldn't make sense to get the second results and therefor it might switch to the minute resolution.

might this be a way to go?

**Comment by alexanderzobnin** [comment]
You can set `Auto` time interval which works similarly.

**Comment by jodok** [comment]
@Node0 we're discussing arbitrary group by intervals. as intermediate workaround, if the grafana query overrides the default limit of 10k by specifying `LIMIT 20000` - would this help?

**Comment by alexanderzobnin** [comment]
This issue appears due database can returns different sets of series. For example, you request CPU load for all servers in group database for the last two hours and use `minute` resolution. This group can contains, for instance, 5 servers or 100 servers. In first case database returns `60 * 2 * 5 = 600` points and `60 * 2 * 100 = 12000` points. And we cannot predict number of series to use corresponding resolution.

**Comment by Node0** [comment]
Hi @jodok!
As an intermediary workaround, I'd say providing a field in the datasource plugin configuration view
where the user can configure the value specified e.g. 10k (default if no value entered), 20k, etc

@alexanderzobnin , I was thinking that a count (*) aggregation statement executed before the main statement-series (I say "statement series" as the count aggregation results would then be able to provide the number of rows necessary to determine how many iterations (each one with a different offset) of the original query statements to run).

Without that count aggregation query, there's no way to know how many rows will need to be fetched. So it becomes essential to first run a modified version of the original query whereby the select clause is replaced by the count(*) clause.

So it's a 'multi-pass' solution, first a count aggregation, which will provide the number of rows necessary to then loop through and request N number of versions of the original query each with an incremented offset (incremented by 10k rows each time).

This solution will handle any number of rows, it's not the fastest solution, however it is stable even when dealing with huge datasets.

To be honest speed in crate is a function of distributed compute, so the speed of the cluster can be increased by scaling. From a data acquisition perspective (from the Grafana side) it seems appropriate to adopt the most reliable way to get ALL the data for a given analytics query, and let the question of response time be left to the configuration of the crate installation (cluster size, replica count, shard count, etc).

Does this approach make sense @alexanderzobnin ?

**Comment by alexanderzobnin** [comment]
@Node0 yes, I think it's stable solution, so we can try this way.

**Comment by jodok** [comment]
@Node0 i agree - that's a really nice workaround right now. @alexanderzobnin let's do it that way and let's get out the beta version of the datasource on grafana.net as soon as possible.

**Comment by joemoe** [comment]
@alexanderzobnin did you work on this issue?

**Comment by alexanderzobnin** [comment]
@joemoe I've added workaround for this issue, so you can set _Group By Time Interval_ to _Auto_ in query editor and check result.

**Comment by Node0** [comment]
Excellent, I'll attempt to test this.
